### ISSUE TITLE
Audio save path is incompatible with TorchAudio 2.9+ and can silently saturate generated WAV files

### ISSUE THREAD

**Issue by John6666cat** [index-tts/index-tts#724]
## Summary

The current IndexTTS audio save path converts the generated normalized waveform to PCM-scale values and then passes an `int16` tensor to `torchaudio.save()`.

This works with the TorchAudio 2.8 save implementation used by the currently supported IndexTTS environment.

However, beginning with TorchAudio 2.9, `torchaudio.save()` delegates encoding to TorchCodec. In that path, PCM-scale `int16` input is converted to `float32` without normalization and is then interpreted as audio that should be in `[-1, 1]`.

The save operation succeeds without an exception or warning, but the resulting WAV is silently saturated. Almost every non-zero sample becomes either negative or positive full scale, destroying the amplitude information.

This report is not claiming that the current official IndexTTS installation is broken. The official dependencies currently pin Torch and TorchAudio to `2.8.*`, which avoids the new save path. The problem appears when this code is used with TorchAudio 2.9 or later, including upgraded, modified, or downstream-packaged environments.

## Current IndexTTS behavior

IndexTTS converts the generated waveform to PCM scale before saving:

```python
wav = torch.clamp(32767 * wav, -32767.0, 32767.0)
```

It later saves that PCM-scale waveform as an integer tensor:

```python
torchaudio.save(
    output_path,
    wav.type(torch.int16),
    sampling_rate,
)
```

This pattern is present in both:

* `indextts/infer.py`
* `indextts/infer_v2.py`

Current source:

https://github.com/index-tts/index-tts/blob/main/indextts/infer.py

https://github.com/index-tts/index-tts/blob/main/indextts/infer_v2.py

The current project dependencies pin:

```toml
torch==2.8.*
torchaudio==2.8.*
```

https://github.com/index-tts/index-tts/blob/main/pyproject.toml

Therefore, the official `uv` environment currently avoids the failure. The save code itself nevertheless becomes incompatible if TorchAudio is updated to 2.9 or later.

## Why it fails

With TorchAudio 2.9 and later, the compatibility wrapper effectively performs:

```python
if src.dtype != torch.float32:
    src = src.float()
```

For example, a PCM int16 value of:

```text
26212
```

becomes:

```text
26212.0
```

rather than a normalized value near:

```text
0.8
```

TorchCodec expects float audio samples in `[-1, 1]`. The out-of-range waveform is accepted and clipped during encoding.

## Minimal reproduction of the IndexTTS save pattern

No IndexTTS model or checkpoint is required to reproduce the save-boundary problem.

```python
from array import array
import sys
import wave

import torch
import torchaudio


sample_rate = 24000
num_samples = sample_rate

t = torch.arange(num_samples, dtype=torch.float32) / sample_rate

# Synthetic normalized model-like output.
normalized = (
    0.72 * torch.sin(2 * torch.pi * 440.0 * t)
    + 0.08 * torch.sin(2 * torch.pi * 997.0 * t)
).clamp(-0.8, 0.8).unsqueeze(0).contiguous()

# Equivalent to the current IndexTTS scaling step.
pcm_scale_wav = torch.clamp(
    32767.0 * normalized,
    -32767.0,
    32767.0,
)

# Current IndexTTS-style save.
torchaudio.save(
    "indextts_current_save.wav",
    pcm_scale_wav.to(torch.int16),
    sample_rate,
)

# Compatible save using normalized float32.
torchaudio.save(
    "indextts_normalized_save.wav",
    (pcm_scale_wav / 32767.0).to(torch.float32),
    sample_rate,
)


def inspect_pcm16(path):
    with wave.open(path, "rb") as wav_file:
        raw = wav_file.readframes(wav_file.getnframes())

    samples = array("h")
    samples.frombytes(raw)

    if sys.byteorder != "little":
        samples.byteswap()

    values = list(samples)

    fullscale_fraction = sum(
        value <= -32768 or value >= 32767
        for value in values
    ) / len(values)

    return {
        "min": min(values),
        "max": max(values),
        "unique_values": len(set(values)),
        "fullscale_fraction": fullscale_fraction,
    }


print(
    "current save:",
    inspect_pcm16("indextts_current_save.wav"),
)

print(
    "normalized save:",
    inspect_pcm16("indextts_normalized_save.wav"),
)
```

## Observed result

Environment:

* Python 3.12.13
* PyTorch 2.11.0+cpu
* TorchAudio 2.11.0+cpu
* TorchCodec 0.15.0+cpu

Current IndexTTS-style save:

```text
{
    'min': -32768,
    'max': 32767,
    'unique_values': 3,
    'fullscale_fraction': 0.999875
}
```

The saved PCM contains effectively only:

```text
-32768, 0, 32767
```

The sign of the waveform is retained, but nearly all amplitude information is lost.

Normalized float32 save:

```text
{
    'min': -26213,
    'max': 26213,
    'unique_values': 18849,
    'fullscale_fraction': 0.0
}
```

## Full reproduction notebook

A public, executable CPU-only notebook is available here:

https://huggingface.co/datasets/John6666/forum3/blob/main/torchaudio_torchcodec_audio_save_reproducer/torchaudio_torchcodec_audio_save_reproducer_final.ipynb

It compares:

* `torchaudio.save()`
* `torchaudio.save_with_torchcodec()`
* direct TorchCodec `AudioEncoder`
* normalized float32 input
* PCM-scale int16 input
* PCM-scale float32 input
* out-of-range float32 input

All automated observations in the notebook evaluate to `OBSERVED`.

## Suggested fix direction

A minimal compatibility fix would be to normalize the existing PCM-scale waveform immediately before file saving:

```python
torchaudio.save(
    output_path,
    (wav.to(torch.float32) / 32767.0).clamp(-1.0, 1.0),
    sampling_rate,
)
```

Because IndexTTS already clamps `wav` to `[-32767, 32767]`, division by `32767.0` restores the expected normalized range.

This approach has several advantages:

* It does not require checking the TorchAudio version.
* It works with the current TorchAudio 2.8 environment.
* It works with the TorchCodec-backed save path in TorchAudio 2.9 and later.
* It does not require changing the existing PCM-scale streaming output.
* It does not require changing the existing int16 NumPy return value used by the UI.

The exact WAV encoding or subtype should remain a maintainer decision. The important requirement is that normalized float samples, rather than PCM-scale integer samples, reach the TorchCodec-backed save path.

A larger future refactor could preserve the normalized model waveform until the final output boundary and create PCM-scale copies only for APIs that specifically require them. That is not necessary for the minimal compatibility fix.

## Upstream issues

TorchAudio compatibility wrapper:

https://github.com/pytorch/audio/issues/4211

TorchCodec value-range validation:

https://github.com/meta-pytorch/torchcodec/issues/1576

The TorchAudio issue covers the legacy integer-input compatibility regression.

The TorchCodec issue separately covers the fact that `AudioEncoder` validates the required dtype but does not reject or warn about float32 values outside the documented `[-1, 1]` range.

## Possibly related IndexTTS reports

The following reports describe loud or heavily corrupted output and may have a similar audible failure mode:

* #540 — Electric and loud static in audio
  https://github.com/index-tts/index-tts/issues/540

* #496 — Noisy output after moving IndexTTS 1.5 and 2.0 to NVIDIA DGX Spark
  https://github.com/index-tts/index-tts/issues/496

Issue #540 reports that every generation sounds similar and very loud.

Issue #496 reports clean output on a Mac mini but unusably noisy output after moving both IndexTTS 1.5 and 2.0 to a different environment.

These reports do not include enough dependency information or pre-save/post-save waveform statistics to establish that they have the same root cause. They are linked only as possibly related audible symptoms.

## Workarounds

For users running TorchAudio 2.9 or later:

```python
torchaudio.save(
    output_path,
    (wav.to(torch.float32) / 32767.0).clamp(-1.0, 1.0),
    sampling_rate,
)
```

Alternatively, using the officially pinned TorchAudio 2.8 environment avoids this specific compatibility failure.

> **AI assistance disclosure:** AI assistance was used to organize and edit this report and the accompanying reproduction notebook. I ran the experiments and reviewed all results and technical claims. The reproduction, observations, and conclusions above reflect my own review.
